ivadomed/testing.py

```diff
@@ -58,21 +59,22 @@ def test(model_params, dataset_test, testing_params, log_directory, device, cuda
     # UNCERTAINTY SETTINGS
     if (testing_params['uncertainty']['epistemic'] or testing_params['uncertainty']['aleatoric']) and \
             testing_params['uncertainty']['n_it'] > 0:
-        n_monteCarlo = testing_params['uncertainty']['n_it']
+        n_monteCarlo = testing_params['uncertainty']['n_it'] + 1
         testing_params['uncertainty']['applied'] = True
-        print('\nComputing model uncertainty over {} iterations.'.format(n_monteCarlo))
+        print('\nComputing model uncertainty over {} iterations.'.format(n_monteCarlo - 1))
     else:
         testing_params['uncertainty']['applied'] = False
         n_monteCarlo = 1
 
     for i_monteCarlo in range(n_monteCarlo):
         preds_npy, gt_npy = run_inference(test_loader, model, model_params, testing_params, path_3Dpred,
-                                          cuda_available, i_monteCarlo)
+                                          cuda_available, i_monteCarlo, postprocessing)
         metric_mgr(preds_npy, gt_npy)
-
-    # COMPUTE UNCERTAINTY MAPS
-    if n_monteCarlo > 1:
-        imed_utils.run_uncertainty(ifolder=path_3Dpred)
+        # If uncertainty computation, don't apply it on last iteration for prediction
+        if testing_params['uncertainty']['applied'] and (n_monteCarlo - 2 == i_monteCarlo):
+            testing_params['uncertainty']['applied'] = False
+            # COMPUTE UNCERTAINTY MAPS
+            imed_utils.run_uncertainty(ifolder=path_3Dpred)
```

Comment:
I am not sure to understand the reason for the extra MC iteration. Could you please help me?

Reply:
With the old code, if there were only 2 MC iterations, `run_inference` would only be run twice and the pred files would be generated in `run_uncertainty`. To avoid code duplication related to applying the postprocessing steps in `run_uncertainty`, an extra iteration is added to go through `run_inference` a last time to generate `_pred.nii.gz` files with postprocessing. With this current code, I think we could remove the code related to the generation of the `_pred.nii.gz` in `run_uncertainty`.